File: solutions/problem_0063_unique_paths_ii/solution.py
"""
Problem 63: Unique Paths II
===========================
Difficulty: Medium
Tags: Array, Dynamic Programming, Matrix
Pattern: Dynamic Programming

Time Complexity:  O(m*n)
Space Complexity: O(n)
"""

class Solution:
    def uniquePathsWithObstacles(self, grid: list[list[int]]) -> int:
        n = len(grid[0])
        dp = [0] * n
        dp[0] = 1
        for row in grid:
            for j in range(n):
                if row[j] == 1:
                    dp[j] = 0
                elif j > 0:
                    dp[j] += dp[j-1]
        return dp[-1]

File: solutions/problem_0063_unique_paths_ii/solution.cpp
/*
 * Problem 63: Unique Paths II
 * ===========================
 * Difficulty: Medium
 * Tags: Array, Dynamic Programming, Matrix
 * Pattern: Dynamic Programming
 *
 * Time Complexity:  O(m*n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        // Dynamic programming (2D) - O(m*n) time and space
        if (obstacleGrid.empty()) return 0;
        int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
            }
        }
        return dp[m][n];
    }
};
